Learning-basics broken

Hi, 

It seems that the Learning-basics module has not been updated to hazelcast-4.0 snapshot.

The learning-basics module configuration files still rely on the 3.11 configuration schema

```
<hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
           xsi:schemaLocation="http://www.hazelcast.com/schema/config
                               http://www.hazelcast.com/schema/config/hazelcast-config-3.11.xsd"
           xmlns="http://www.hazelcast.com/schema/config">
[...]
</hazelcast>
```

Running the learning examples as-is leads to an XML exception (considering that the schemas don't match).
Upgrading the versions in the XML files does not work as there seems to be some breaking changes in the configuration schema.

- Downgrading the learning basics module to version 3.11 in the POM leads to another exception being thrown.
```
Caused by: java.lang.IllegalStateException: Unable to get `original` field in `HazelcastInstanceProxy`!
	at com.hazelcast.examples.helper.CommonUtils.<clinit>(CommonUtils.java:49)
	... 1 more
Caused by: java.lang.NoClassDefFoundError: com/hazelcast/instance/impl/HazelcastInstanceProxy
	at com.hazelcast.examples.helper.CommonUtils.<clinit>(CommonUtils.java:46)
	... 1 more
Caused by: java.lang.ClassNotFoundException: com.hazelcast.instance.impl.HazelcastInstanceProxy
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 2 more
```

In the meantime I guess I'll check out a previous commit that still uses version 3.XX to get familiar with Hazelcast.

--- a/distributed-map/eviction/src/main/java/MapCustomEvictionPolicy.java
+++ b/distributed-map/eviction/src/main/java/MapCustomEvictionPolicy.java
@@ -22,8 +22,9 @@ import com.hazelcast.core.EntryView;
 import com.hazelcast.core.Hazelcast;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.map.IMap;
-import com.hazelcast.map.eviction.MapEvictionPolicy;
+import com.hazelcast.map.MapEvictionPolicyComparator;
 import com.hazelcast.map.listener.EntryEvictedListener;
+import com.hazelcast.spi.eviction.EvictionPolicyComparator;
 
 import java.util.Queue;
 import java.util.concurrent.ConcurrentLinkedQueue;
@@ -34,21 +35,22 @@ import static java.util.concurrent.TimeUnit.SECONDS;
 import static java.util.concurrent.locks.LockSupport.parkNanos;
 
 /**
- * This sample shows custom-eviction-policy usage for IMap. IMap uses a sampling based approach to evict entries.
+ * This sample shows custom eviction policy usage for IMap. IMap uses a sampling based approach to evict entries.
  * For more info about sampling approach, please refer to hazelcast documentation.
  *
- * To plug a custom-eviction-policy {@link com.hazelcast.config.MapConfig#setMapEvictionPolicy(MapEvictionPolicy)} is used
- * in this example, it is also pluggable declaratively.
+ * To plug a custom eviction policy
+ * {@link com.hazelcast.config.EvictionConfig#setComparator(EvictionPolicyComparator)} is used
+ * in this example. Alternatively, custom eviction policy's class name can be also configured declaratively.
  */
 public class MapCustomEvictionPolicy {
 
     public static void main(String[] args) {
         Config config = new Config();
         MapConfig mapConfig = config.getMapConfig("test");
-        mapConfig.setMapEvictionPolicy(new OddEvictor());
         mapConfig.getEvictionConfig()
                 .setMaxSizePolicy(MaxSizePolicy.PER_NODE)
-                .setSize(10000);
+                .setSize(10000)
+                .setComparator(new OddEvictor());
 
         HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
         IMap<Integer, Integer> map = instance.getMap("test");
@@ -86,11 +88,11 @@ public class MapCustomEvictionPolicy {
     /**
      * Odd evictor tries to evict odd keys first.
      */
-    private static class OddEvictor extends MapEvictionPolicy {
+    private static class OddEvictor implements MapEvictionPolicyComparator<Integer, Integer> {
 
         @Override
-        public int compare(EntryView o1, EntryView o2) {
-            Integer key = (Integer) o1.getKey();
+        public int compare(EntryView<Integer, Integer> o1, EntryView<Integer, Integer> o2) {
+            Integer key = o1.getKey();
             if (key % 2 != 0) {
                 return -1;
             }

--- a/jcache/src/main/java/com/hazelcast/examples/eviction/CacheCustomEvictionPolicyUsage.java
+++ b/jcache/src/main/java/com/hazelcast/examples/eviction/CacheCustomEvictionPolicyUsage.java
@@ -28,7 +28,9 @@ public class CacheCustomEvictionPolicyUsage {
 
         // Use eviction policy comparator by its instance
         CacheConfig cacheConfig1 = createCacheConfig("MyCache_1", false);
-        ICache<Integer, String> cache1 = (ICache<Integer, String>) cacheManager.createCache("MyCache_1", cacheConfig1);
+        ICache<Integer, String> cache1 = (ICache<Integer, String>) cacheManager
+                .createCache("MyCache_1", cacheConfig1)
+                .unwrap(ICache.class);
 
         // Default max-size for eviction is 10K.
         // With partition based cluster wide approximation, after almost 15K elements (for default 271 partition),
@@ -39,7 +41,9 @@ public class CacheCustomEvictionPolicyUsage {
 
         // Use eviction policy comparator by its name
         CacheConfig cacheConfig2 = createCacheConfig("MyCache_2", true);
-        ICache<Integer, String> cache2 = (ICache<Integer, String>) cacheManager.createCache("MyCache_2", cacheConfig2);
+        ICache<Integer, String> cache2 = (ICache<Integer, String>) cacheManager
+                .createCache("MyCache_2", cacheConfig2)
+                .unwrap(ICache.class);
 
         // Default max-size for eviction is 10K.
         // With partition based cluster wide approximation, after almost 15K elements (for default 271 partition),
@@ -60,7 +64,7 @@ public class CacheCustomEvictionPolicyUsage {
                 .setEvictionConfig(evictionConfig);
     }
 
-    private static class MyEvictionPolicyComparator extends CacheEvictionPolicyComparator<Integer, String> {
+    private static class MyEvictionPolicyComparator implements CacheEvictionPolicyComparator<Integer, String> {
 
         @Override
         public int compare(CacheEntryView<Integer, String> e1, CacheEntryView<Integer, String> e2) {
@@ -72,13 +76,13 @@ public class CacheCustomEvictionPolicyUsage {
             }
             if (key2 > key1) {
                 // -1
-                return FIRST_ENTRY_HAS_HIGHER_PRIORITY_TO_BE_EVICTED;
+                return -1;
             } else if (key2 < key1) {
                 // +1
-                return SECOND_ENTRY_HAS_HIGHER_PRIORITY_TO_BE_EVICTED;
+                return 1;
             } else {
                 // 0
-                return BOTH_OF_ENTRIES_HAVE_SAME_PRIORITY_TO_BE_EVICTED;
+                return 0;
             }
         }
     }

--- a/learning-basics/configure-logging/src/main/resources/hazelcast.xml
+++ b/learning-basics/configure-logging/src/main/resources/hazelcast.xml
@@ -1,6 +1,6 @@
 <hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-                               http://www.hazelcast.com/schema/config/hazelcast-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd"
            xmlns="http://www.hazelcast.com/schema/config">
 
     <network>

--- a/learning-basics/configure-xml/src/main/resources/hazelcast-client-network-config.xml
+++ b/learning-basics/configure-xml/src/main/resources/hazelcast-client-network-config.xml
@@ -17,7 +17,7 @@
 
 <hazelcast-client xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xsi:schemaLocation="http://www.hazelcast.com/schema/client-config
-                               http://www.hazelcast.com/schema/client-config/hazelcast-client-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/client-config/hazelcast-client-config-4.0.xsd"
                   xmlns="http://www.hazelcast.com/schema/client-config">
 
     <network>

--- a/learning-basics/configure-xml/src/main/resources/hazelcast-client.xml
+++ b/learning-basics/configure-xml/src/main/resources/hazelcast-client.xml
@@ -17,12 +17,9 @@
 
 <hazelcast-client xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xsi:schemaLocation="http://www.hazelcast.com/schema/client-config
-                               http://www.hazelcast.com/schema/client-config/hazelcast-client-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/client-config/hazelcast-client-config-4.0.xsd"
                   xmlns="http://www.hazelcast.com/schema/client-config">
 
     <import resource="hazelcast-client-network-config.xml"/>
-    <group>
-        <name>dev-configured-from-xml</name>
-        <password>1234</password>
-    </group>
+    <cluster-name>dev-configured-from-xml</cluster-name>
 </hazelcast-client>

--- a/learning-basics/configure-xml/src/main/resources/hazelcast-network-config.xml
+++ b/learning-basics/configure-xml/src/main/resources/hazelcast-network-config.xml
@@ -1,6 +1,6 @@
 <hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-                               http://www.hazelcast.com/schema/config/hazelcast-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd"
            xmlns="http://www.hazelcast.com/schema/config">
 
     <network>

--- a/learning-basics/configure-xml/src/main/resources/hazelcast.xml
+++ b/learning-basics/configure-xml/src/main/resources/hazelcast.xml
@@ -1,11 +1,8 @@
 <hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-                               http://www.hazelcast.com/schema/config/hazelcast-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd"
            xmlns="http://www.hazelcast.com/schema/config">
 
     <import resource="hazelcast-network-config.xml"/>
-    <group>
-        <name>dev-configured-from-xml</name>
-        <password>1234</password>
-    </group>
+    <cluster-name>dev-configured-from-xml</cluster-name>
 </hazelcast>

--- a/learning-basics/configure-yaml/src/main/resources/hazelcast-client.yaml
+++ b/learning-basics/configure-yaml/src/main/resources/hazelcast-client.yaml
@@ -1,6 +1,4 @@
 hazelcast-client:
   import:
     - hazelcast-client-network-config.yaml
-  group:
-    name: dev-configured-from-yaml
-    password: 1234
+  cluster-name: dev-configured-from-yaml

--- a/learning-basics/configure-yaml/src/main/resources/hazelcast.xml
+++ b/learning-basics/configure-yaml/src/main/resources/hazelcast.xml
@@ -1,6 +1,6 @@
 <hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-                               http://www.hazelcast.com/schema/config/hazelcast-config-3.12.xsd"
+                               http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd"
            xmlns="http://www.hazelcast.com/schema/config">
     <network>
         <port>6000</port>

--- a/learning-basics/configure-yaml/src/main/resources/hazelcast.yaml
+++ b/learning-basics/configure-yaml/src/main/resources/hazelcast.yaml
@@ -1,6 +1,4 @@
 hazelcast:
   import:
     - hazelcast-network-config.yaml
-  group:
-    name: dev-configured-from-yaml
-    password: 1234
+  cluster-name: dev-configured-from-yaml

--- a/learning-basics/destroying-instances/src/main/resources/hazelcast.xml
+++ b/learning-basics/destroying-instances/src/main/resources/hazelcast.xml
@@ -1,6 +1,6 @@
 <hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-                               http://www.hazelcast.com/schema/config/hazelcast-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd"
            xmlns="http://www.hazelcast.com/schema/config">
 
     <network>

--- a/learning-basics/loading-instances/src/main/resources/hazelcast.xml
+++ b/learning-basics/loading-instances/src/main/resources/hazelcast.xml
@@ -1,6 +1,6 @@
 <hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-                               http://www.hazelcast.com/schema/config/hazelcast-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd"
            xmlns="http://www.hazelcast.com/schema/config">
 
     <queue name="q"/>

--- a/learning-basics/wildcard-configuration/src/main/java/WildcardMember.java
+++ b/learning-basics/wildcard-configuration/src/main/java/WildcardMember.java
@@ -13,8 +13,8 @@ public class WildcardMember {
         map1.put("foo", "foo");
         map2.put("foo", "foo");
         for (; ; ) {
-            System.out.println("size:" + map1.size());
-            System.out.println("size2:" + map2.size());
+            System.out.println("size of map1:" + map1.size());
+            System.out.println("size of map2:" + map2.size());
             Thread.sleep(1000);
         }
     }

--- a/learning-basics/wildcard-configuration/src/main/resources/hazelcast.xml
+++ b/learning-basics/wildcard-configuration/src/main/resources/hazelcast.xml
@@ -1,6 +1,6 @@
 <hazelcast xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-                               http://www.hazelcast.com/schema/config/hazelcast-config-3.11.xsd"
+                               http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd"
            xmlns="http://www.hazelcast.com/schema/config">
 
     <map name="testmap*">

